html/modules/custom/reliefweb_import/src/Plugin/ReliefWebImporter/FaoDataImporter.php

```diff
@@ -0,0 +1,466 @@
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state): array {
+    $form = parent::buildConfigurationForm($form, $form_state);
+
+    $form['api_url'] = [
+      '#type' => 'url',
+      '#title' => $this->t('API URL'),
+      '#description' => $this->t('The base URL of the FAO Data API.'),
+      '#default_value' => $form_state->getValue('api_url', $this->getPluginSetting('api_url', '', FALSE)),
+      '#required' => TRUE,
+    ];
+
+    $form['timeout'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Timeout'),
+      '#description' => $this->t('Connection and request timeout in seconds.'),
+      '#default_value' => $form_state->getValue('timeout', $this->getPluginSetting('timeout', 5, FALSE)),
+      '#min' => 1,
+      '#required' => TRUE,
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function importContent(int $limit = 50): bool {
+    // Get list of documents.
+    try {
+      $provider_uuid = $this->getPluginSetting('provider_uuid');
+
+      // Retrieve the POST API content processor plugin.
+      $plugin = $this->contentProcessorPluginManager->getPluginByResource('reports');
+
+      // Ensure the provider is valid.
+      $plugin->getProvider($provider_uuid);
+
+      $this->getLogger()->info('Retrieving documents from the FAO data API.');
+
+      // Retrieve the latest created documents.
+      $documents = $this->getDocuments($limit);
+
+      if (empty($documents)) {
+        $this->getLogger()->notice('No documents.');
+        return TRUE;
+      }
+    }
+    catch (\Exception $exception) {
+      $this->getLogger()->error($exception->getMessage());
+      return FALSE;
+    }
+
+    $this->getLogger()->info(strtr('Retrieved @count FAO documents.', [
+      '@count' => count($documents),
+    ]));
+
+    // Sort the documents by ID ascending to process the oldest ones first.
+    ksort($documents);
+
+    // Process the documents importing new ones and updated ones.
+    $processed = $this->processDocuments($documents, $provider_uuid, $plugin);
+
+    // @todo check if we want to return TRUE only if there was no errors or if
+    // return TRUE for partial success is fine enough.
+    return $processed > 0;
+  }
+
+  /**
+   * Retrieve documents from the FAO API.
+   *
+   * @param int $limit
+   *   Maximum number of documents to retrieve at once.
+   * @param string $order_property
+   *   Property to use to sort the documents.
+   *
+   * @return array
+   *   List of documents keyed by IDs.
+   */
+  protected function getDocuments(int $limit, string $order_property = 'created'): array {
+    // Get list of documents.
+    try {
+      $timeout = $this->getPluginSetting('timeout', 5, FALSE);
+      $api_url = $this->getPluginSetting('api_url');
+
+      // Query the FAO API.
+      $query = http_build_query([
+        'limit' => $limit,
+        'type' => 'PDF',
```

Comment:
Looks like we could use something like `filter=(type=PDF AND (created BETWEEN 1745938800000 AND 1750517999999))` to filter for PDF and restrict the date range. So we could store the timestamp of the latest import and use that for the lower time limit to reduce import attempts of the same content. What do you think?

Reply:
:confused: code updated, will generate a link like `https://data-in-emergencies.fao.org/api/search/v1/collections/document/items?limit=5&sortBy=%2Bproperties.created&filter=%28type%3DPDF+AND+%28created+BETWEEN+1650168396+AND+1750754796%29%29` but this doesn't return anything

I tested on their site as well, but filter doesn't apply

Reply:
The timestamps need to be in milliseconds. `https://data-in-emergencies.fao.org/api/search/v1/collections/document/items?limit=5&sortBy=%2Bproperties.created&filter=(type%3DPDF+AND+(created+BETWEEN+1650168396000+AND+1750754796000))` seems to work.

Reply:
Seems I needed to go back in time a bit more

:+1: